Add ability to overwrite the Kubernetes url and SSL parameters

aerostitch opened the pull request:
This does a little more than what was requested in #275 but I need to be able to use a custom url if I want to use my custom ca file so I figured I could add the `verify_ssl` at the same time.

Closes #275

repeatedly commented on templates/conf/kubernetes.conf.erb:
```diff
@@ -258,6 +258,9 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['KUBERNETES_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
```
"KUBERNETES_" prefix seems general.
How about "FLUENTD_FILTER_KUBERNETS_URL" or something?
Other environment variable too.

aerostitch replied:
For `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` this is referring to the standard variables that kubernetes makes available inside each pod but I guess you know that already.
I put them here to match the default https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter/blob/master/lib/fluent/plugin/filter_kubernetes_metadata.rb#L195 as we're specifying `kubernetes_url` here in any case.

I changed `KUBERNETES_URL` to `FLUENT_FILTER_KUBERNETS_URL`, rebased from master re-ran `make src-all`.

idohalevi commented:
@repeatedly 
I need this too, when do you think you will merge?

ganmacs commented on docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf:
```diff
@@ -188,5 +188,8 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
```
I think this won't work.
How do you feel like the following code?
` "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || "https://#{ENV.fetch('KUBERNETES_SERVICE_HOST')}:#{ENV.fetch('KUBERNETES_SERVICE_PORT')}/api"}"`

aerostitch replied:
That's fine with me too. Updated the PR.

aerostitch commented on docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf:
```diff
@@ -188,5 +188,8 @@
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
```
Good catch, thanks!

aerostitch commented on docker-image/v1.3/debian-cloudwatch/conf/kubernetes.conf:
```diff
@@ -188,5 +188,8 @@
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
+  ca_file "#{ENV['KUBERNETES_CA_FILE'] || ''}"
```
Indeed, updated.

aerostitch commented:
Sorry for the late update I missed the latest notifications.
I updated the PR based on @repeatedly and @ganmacs 's feedbacks.
Let me know if you need me to do other changes.

Thanks!

Timoses commented:
Why is this not in yet? It seems like a crucial part to get fluentd Kubernetes running on Clusters which use a custom CA for the Kubernetes API.

ganmacs commented on docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf:
```diff
@@ -188,5 +188,8 @@
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
```
https://github.com/fluent/fluentd-kubernetes-daemonset/pull/279/files/532befe36d84250a8d50b8cc8aeeaccd61206b1d#diff-bf350a335c264690ed81e269cca50f23

All others are the same.

aerostitch replied:
@ganmacs I don't understand your comment. Could you be more explicit in what change you are looking for please?

aerostitch replied:
Never mind, I finally understood! Sorry.

aerostitch replied:
Updated

ganmacs commented on docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf:
```diff
@@ -188,5 +188,8 @@
 <filter kubernetes.**>
   @type kubernetes_metadata
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
```
https://github.com/fluent/fluentd-kubernetes-daemonset/pull/279/files/532befe36d84250a8d50b8cc8aeeaccd61206b1d#diff-bf350a335c264690ed81e269cca50f23

All others are the same.

aerostitch replied:
Sorry for the late reply.
@ganmacs I don't understand your comment. Could you be more explicit in what change you are looking for please?

aerostitch replied:
Never mind, I finally understood! Sorry.

ganmacs commented on docker-image/v1.3/debian-elasticsearch/conf/kubernetes.conf:
```diff
@@ -188,5 +188,8 @@
   @id filter_kube_metadata
+  kubernetes_url "#{ENV['FLUENT_FILTER_KUBERNETS_URL'] || https://#{ENV['KUBERNETES_SERVICE_HOST']}:#{ENV['KUBERNETES_SERVICE_PORT']}/api}"
+  verify_ssl #{ENV['KUBERNETES_VERIFY_SSL'] || true}
+  ca_file "#{ENV['KUBERNETES_CA_FILE'] || ''}"
```
https://github.com/fluent/fluentd-kubernetes-daemonset/pull/279/files/532befe36d84250a8d50b8cc8aeeaccd61206b1d#diff-bf350a335c264690ed81e269cca50f23

All others are the same

aerostitch replied:
@ganmacs I don't understand your comment. Could you be more explicit in what change you are looking for please?

aerostitch replied:
Never mind, I finally understood! Sorry.

ganmacs commented:
Sorry for the delay. I added some comments.
And if possible, could you please write a more descriptive commit message... ("Update with the feedback from ganmacs" doesn't make sense to other developers... )

aerostitch commented:
Sorry for the late reply.
There are some comments I really don't understand what you meant. Could you develop please?
For the commit message aren't you going to use "Squash and merge" when merging in master to avoid having several commits?

aerostitch commented:
Hi,

FYI I rebased the branch against the latest master, fixed a typo in a variable name, updated the template based on the remaining unaddressed feedback and regenerated the images for version 1.3, 1.4 and 1.7 (current) as this PR was addressing 1.3 when it was originally opened.
Let me know if there are still any blocker to get this merge please.

Thanks,
Joseph

repeatedly commented:
Thanks for working on this.
I will release new image soon.

aerostitch commented:
Awesome, thanks a lot! :)
